pkg/collector/corechecks/servicediscovery/impl_linux.go

```diff
@@ -0,0 +1,312 @@
+	events.stopped = append(events.stopped, svc)
+}
+
+func (li *linuxImpl) DiscoverServices() error {
+	procs, err := li.aliveProcs()
+	if err != nil {
+		return fmt.Errorf("failed to get alive processes: %w", err)
+	}
+
+	ports, err := li.openPorts()
+	if err != nil {
+		return fmt.Errorf("failed to get open ports: %w", err)
+	}
+
+	log.Debugf("aliveProcs: %d | ignoreProcs: %d | runningServices: %d | potentials: %d | openPorts: %+v",
+		len(procs),
+		len(li.ignoreProcs),
+		len(li.aliveServices),
+		len(li.potentialServices),
+		ports,
+	)
+
+	var (
+		started []*serviceInfo
+		stopped []*serviceInfo
+	)
+
+	// potentialServices contains processes that we scanned in the previous iteration and had open ports.
+	// we check if they are still alive in this iteration, and if so, we send a start-service telemetry event.
+	for pid, _ := range li.potentialServices {
+		if p, ok := procs[pid]; ok {
+			svc, err := li.getServiceInfo(p, ports)
+			if err != nil {
+				log.Errorf("[pid: %d] failed to get service info: %v", pid, err)
+				continue
+			}
+			li.aliveServices[pid] = svc
+			started = append(started, svc)
+		}
+	}
+	clear(li.potentialServices)
+
+	// check open ports - these will be potential new services if they are still alive in the next iteration.
+	for pid, _ := range ports {
+		if li.ignoreProcs[pid] {
+			continue
+		}
+		if _, ok := li.aliveServices[pid]; !ok {
+			log.Debugf("[pid: %d] found new process with open ports: %+v", pid, li.aliveServices)
+
+			p := procs[pid]
+			svc, err := li.getServiceInfo(p, ports)
+			if err != nil {
+				log.Errorf("[pid: %d] failed to get process info: %v", pid, err)
+				continue
+			}
+			if li.ignoreCfg[svc.meta.Name] {
+				log.Debugf("[pid: %d] process ignored from config: %s", pid, svc.meta.Name)
+				li.ignoreProcs[pid] = true
+				continue
+			}
+			log.Debugf("[pid: %d] adding process to potential: %s", pid, svc.meta.Name)
+			li.potentialServices[pid] = svc
+		}
+	}
+
+	// check if services previously marked as alive still are.
+	now := li.time.Now()
+	for pid, svc := range li.aliveServices {
+		if _, ok := procs[pid]; !ok {
+			delete(li.aliveServices, pid)
+			stopped = append(stopped, svc)
+		} else if now.Sub(svc.LastHeartbeat).Truncate(time.Minute) >= heartbeatTime {
+			li.sender.sendHeartbeatServiceEvent(svc)
+			svc.LastHeartbeat = now
+		}
+	}
+
+	// check if services previously marked as ignore are still alive.
+	for pid, _ := range li.ignoreProcs {
+		if _, ok := procs[pid]; !ok {
+			delete(li.ignoreProcs, pid)
+		}
+	}
+
+	// group started and stopped processes by name
+	events := make(eventsByName)
+	for _, p := range started {
+		events.addStarted(p)
+	}
+	for _, p := range stopped {
+		events.addStopped(p)
+	}
+
+	for name, ev := range events {
+		if len(ev.started) > 0 && len(ev.stopped) > 0 {
+			// TODO: do something smarter here
+			log.Warnf("found multiple started/stopped processes with the same name, skipping events (name: %q)", name)
```

Comment:
we should still send the start service events and ignore the end service events. That way we don't time out the row in REDAPL and the start time is kept up to date.